Control/DeepSeq.hs

```diff
@@ -179,6 +179,13 @@ instance GNFData arity a => GNFData arity (M1 i c a) where
   grnf args = grnf args . unM1
   {-# INLINEABLE grnf #-}
 
+#if __GLASGOW_HASKELL__ >= 800
+  -- @since 1.4.5.0
+instance GNFData Zero (URec a) where
+  grnf _ _ = ()
```

Comment:
This doesn't seem right, as this doesn't actually evaluate the thing of type `URec a`. I think you actually want the implementation to be this:

```hs
  grnf _ = rwhnf
```

Of course, this assumes that for every `URec a`, we have that WHNF = NF. This should always be the case, as every `URec` data instance is a datatype with a single strict field. Still, it would be worth leaving a brief comment stating why `rwhnf` is the right choice here.

Reply:
It doesn't evaluate the unboxed values because they're already evaluated! These types simply don't contain unevaluated computations so there is nothing we need to do.

Reply:
I'm still not convinced that this is right. Here is a program which illustrates my concern:

```hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE MagicHash #-}
module Main where

import Control.DeepSeq
import GHC.Exts (Int#)
import GHC.Generics

data Blah = MkBlah Int#
  deriving Generic

instance NFData Blah

b :: Blah
b = undefined

main :: IO ()
main = b `deepseq` pure ()
```

Startingly, this program does _not_ throw an `undefined` exception when ran, which seems very wrong.

Reply:
Thanks! I clearly didn't understand how GNFData works.

Reply:
i just tried this after you made the recommended change, and it throws an undefined exception, as expected. marking resolved